### DISCUSSION TITLE
Apply same parameter to all routes

### DISCUSSION THREAD

**Discussion by xiki808** [tighten/ziggy#615] [Q&A]
#### Description

I need to pass the same parameter to all my routes. I don't find any callback available while registering Ziggy to attach such a parameter to all routes.

#### Suggestion

A callback function on setup to modify each route as needed.

**Comment by bakerkretzmar** [comment]
You can do this on the Laravel side and Ziggy will pick it up—call `URL::defaults()` in a service provider.

Docs: https://laravel.com/docs/9.x/urls#default-values

**Reply by xiki808** [reply]
Thank you, will check it!

**Reply by xiki808** [reply]
Hi @bakerkretzmar, thank you for your tip, but it's not exactly what I need. I had `URL::defaults()` already setup, but this only helps with setting a default param to the route on the backend. In my case I would like to have a default param set from the frontend side. Currently I have to add it manually for each route. Would be perfect if there's a callback function that adds such a param to all routes from one place.

**Reply by bakerkretzmar** [reply]
Ziggy uses `URL::defaults()` on the frontend too. Are you using the `@routes` Blade directive? If you are, and you call `URL::defaults(['locale' => 'en'])` before `@routes` runs, you can leave the `locale` parameter out of all your `route()` calls and it will be populated with `en` automatically.

**Reply by xiki808** [reply]
No I'm using vue 3 instead of Blade, and I have it setup like so. 

```
createInertiaApp({
  title: title => `${title} - ${appName}`,
  resolve: name =>
    resolvePageComponent(
      `./Pages/${name}.vue`,
      import.meta.glob('./Pages/**/*.vue')
    ),
  setup ({ el, app, props, plugin }) {
    return createApp({ render: () => h(app, props) })
      .use(plugin)
      .use(createPinia())
      .use(ZiggyVue)
  },
});
```

What I really want is to overwrite the locale variable on each route click from the frontend, to do so I am manually including the locale as a parameter on each. I need this, because when I switch the language, locale in html lang and browser url changes, but all the other routes in the DOM remain the same. For example, if I switch to French, the logo link is still http://localhost/en. To fix that I manually pass the locale parameter, with a value of the current language. Would be great if i can do this only in one place and not for all the routes

**Reply by bakerkretzmar** [reply]
Oh gotcha. Take a look at #465—I think you can get what you want but it'll require a bit more manual setup.

Instead of generating a `ziggy.js` file if you're doing that, you'll need to load your Ziggy config on every request and pass it back with Inertia's shared data. Then, instead of calling `URL::defaults(['locale' => 'en'])`, do `URL::defaults(['locale' => app()->getLocale()])` (or equivalent), so that every time your routes are generated the `locale` param defaults to the current locale at that moment.

That way you should only need to explicitly pass `locale` when you want to _change_ it, otherwise it'll default to the current locale.

Let me know if that makes sense and if you get it working, otherwise I can try to put together a demo repo to show you what I mean.

**Reply by xiki808** [reply]
I might have set things up wrongly, but the way I have it, I need a locale parameter passed in every route, since I am using that to prefix my routing on the backend, like so:

```
Route::prefix('{locale}')
    ->where(['locale' => '[a-zA-Z]{2}'])
    ->middleware('setlocale')
    ->group(function () {
    .....
```

This setup I followed from: https://laraveldaily.com/post/multi-language-routes-and-locales-with-auth

If you have any suggestion to improve this, I would like to know.

**Reply by bakerkretzmar** [reply]
Try this and let me know if it works: https://github.com/tighten/ziggy/discussions/463#discussioncomment-1381011

That, combined with `URL::defaults(['locale' => app()->getLocale()])`, should ensure you get a Ziggy payload on the frontend that always has a default parameter for `locale` set to the current locale.

Note that you will need to call `URL::defaults()` **before** the Inertia sharing code runs, e.g. in an earlier middleware.

**Reply by xiki808** [reply]
Thanks for your reply. I believe I have the locale in the Ziggy payload already:

![image](https://user-images.githubusercontent.com/8024321/216982843-1d45745f-4683-41b8-9ff7-d8080432e14f.png)

From my translation library in the frontend I'm using a function to get the the current language. This I'm using to pass the current language to each route. By doing so, on language change all routes are patched and the language segment part of all links are updated. Removing the language ( locale ) parameter from the route, won't cause the links to rerender with the newly selected language. I get a weird behaviour then, the page clicked will be translated accordingly to the current route, but on the browser address bar, I still see the previous language.

**Reply by bakerkretzmar** [reply]
Ahh gotcha. Yeah if switching the language doesn't reload the page then you would have to either regenerate all the current page's links manually or refactor so that the URLs are all computed or something and regenerate automatically when the locale changes.
